include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -902,124 +842,265 @@ void EndParallelSectionInternal(PerThread &pt,
+//   the parallel section, as opposed to just a single loop's
+//   duration).
+//
+// For ordinary parallel sections, RunInParallelInternal dispatch
+// tasks to a number of workers asynchronously.  A worker thread will
+// be selected as the dispatcher that distributes tasks.  This removes
+// the O(n) work off the critical path of starting the first loop
+// iteration, helping maintain good performance on very short loops.
+
 void RunInParallelInternal(PerThread& pt,
                            ThreadPoolParallelSection& ps,
-                           unsigned n,
+                           unsigned new_dop,
+                           bool dispatch_async,
                            std::function<void(unsigned)> worker_fn) {
-  // init a few env variables
-  ps.dispatch_q_idx = -1;
-  ps.dispatch_done = false;
-  ps.work_done = false;
-  // calculate how many workers to summon
-  unsigned current_dop = static_cast<unsigned>(ps.tasks.size()) + 1;
-  unsigned extra_needed = n > current_dop ? n - current_dop : 0;
-  if (0 == extra_needed) {
-    return; // do nothing if no more workers required
-  }
-  // get workers that are available
-  GetGoodWorkerHints(extra_needed, ps.good_hints, ps.alt_hints);
-  // define task for each worker
-  Task call_worker_fn = [worker_fn, &ps]() {
-    worker_fn(++ps.worker_idx);
-    ps.tasks_finished++;
-  };
-  // define dispatch task for dispatcher
-  Task dispatch_task = [extra_needed, call_worker_fn, worker_fn, &ps, &pt, this]() {
-    unsigned good_hints_size = static_cast<unsigned>(ps.good_hints.size());
-    for (auto i = 1u; i < extra_needed; ++i) {  // dispatcher do dispatch
-      int q_idx;
-      if (i < ps.good_hints.size()) {
-        q_idx = ps.good_hints[i];
-      } else {
-        auto alt_i = i - good_hints_size;
-        if (alt_i < ps.alt_hints.size()) {
-          q_idx = ps.alt_hints[alt_i];
-        } else {
-          q_idx = Rand(&pt.rand) % num_threads_;
+  assert(new_dop <= (unsigned)(num_threads_+1));
+  std::vector<int> &preferred_workers = pt.preferred_workers;
+  InitializePreferredWorkers(preferred_workers);
+
+  // current_dop is the degree of parallelism via any workers already
+  // participating in the current parallel section.  Usually, for
+  // single-loop parallel sections, current_dop=1.
+  unsigned current_dop = ps.current_dop;
+  
+  if (current_dop < new_dop) {
+    unsigned extra_needed = new_dop - current_dop;
+
+    // Attempt to summon additional workers asynchronously if we
+    // need more than one.  Otherwise, we fall back to simple
+    // synchronous scheduling.
+    if (dispatch_async && extra_needed > 1) {
+      assert(current_dop == 1);
+      
+      // Task for dispatching work asynchronously.
+      Task dispatch_task = [current_dop, new_dop, worker_fn, &preferred_workers, &ps, &pt, this]() {
+        // Schedule tasks par_idx=[current_dop+1,new_dop)
+        ps.dispatch_started.store(true, std::memory_order_seq_cst);
```

Comment:
>dispatch_started [](http://example.com/codeflow?start=11&length=16)

Even that dispatch_task is already started, there is a chance that "dispatch_started" stays false for a while. So why we need it?

Reply:
I'll expand the comment on the logic here.